### DISCUSSION TITLE
How to use StateFlow instead of Value?

### DISCUSSION THREAD

**Discussion by mr-kew** [arkivanov/Decompose#816] [Q&A]
Hello,

I wanna combine multiplatform decompse combined with [KMP-ObservableViewModel](https://github.com/rickclephas/KMP-ObservableViewModel) library. The advantage should be that I won't have to implement the constructor manually in each view. Also my models are already done using that library, so I would prefer not having to rework them all.

And for that I would wanna use `StateFlow` instead of `Value` for the child stack or child slot. What is the best way to do it? I came up with the following code, but I am not sure if doing local `MutableStateFlow` is ok.

```
abstract class ObservableComponent(
    componentContext: ComponentContext,
) : ViewModel(), ComponentContext by componentContext {

    protected inline fun <reified T : Any> Value<T>.toStateFlow(mode: ObserveLifecycleMode = ObserveLifecycleMode.START_STOP): StateFlow<T> {
        val state = MutableStateFlow(viewModelScope, this.value)
        this.subscribe(lifecycle, mode) { value ->
            state.value = value
        }
        return state.asStateFlow()
    }
}
```

Usage:
(of course the `toStateFlow` could be hidden inside a `childSlot` overload)
```
class SlotComponent(
    componentContext: ComponentContext,
) : ObservableComponent(componentContext) {
    private val navigation = SlotNavigation<Config>()

    @NativeCoroutinesState
    val slot: StateFlow<ChildSlot<*, Child>> = childSlot(
        source = navigation,
        serializer = Config.serializer(),
        initialConfiguration = { null },
        handleBackButton = true,
        childFactory = ::child,
    ).toStateFlow()

    ...
```

**Comment by CXwudi** [comment]
This might help?

https://gist.github.com/CXwudi/2c4084eb0e82c05fd943b040fc73c2f9#file-decomposevalueutil-kt-L72-L85

**Comment by arkivanov** [comment]
Oops, seems like I missed this question. Sorry about that.
In addition to what was suggest before, there is also the following way:

```kotlin
import com.arkivanov.decompose.value.Value
import kotlinx.coroutines.flow.FlowCollector
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

fun <T : Any> Value<T>.asStateFlow(): StateFlow<T> =
    ValueStateFlow(this)

private class ValueStateFlow<out T : Any>(
    private val v: Value<T>,
) : StateFlow<T> {

    override val value: T get() = v.value
    override val replayCache: List<T> get() = listOf(v.value)

    override suspend fun collect(collector: FlowCollector<T>): Nothing {
        val stateFlow = MutableStateFlow(v.value)
        val cancellation = v.subscribe { stateFlow.value = it }

        try {
            stateFlow.collect(collector)
        } finally {
            cancellation.cancel()
        }
    }
}
```

**Reply by YektaDev** [reply]
Hi! I hate to nitpick, but as someone who's all in on `StateFlow`s in all Components, this _feels_ too much for the sole purpose of holding a value. By that, I mean the sum cost of below items per `ValueStateFlow`:
- The locking mechanism inside `Value` + `StateFlow`s suspending `collect`.
- Object creations: Value internals (`MutableValue`, `Lock`, `observers` map) + Mapper internals (`replayCache` list, `MutableStateFlow` in `collect`) + StateFlow internals (I have no idea)

This is merely a thought that I felt the need to share. I know it might be a negligible cost in the grand scheme of things, and I'm ok with the current solution as I can't think of a better alternative (that's without wishing the `Value` didn't exist in the first place😃).

**Reply by arkivanov** [reply]
Thanks for sharing! The overhead is indeed negligible, taking into account that `asStateFlow` is supposed to be called only once per component per `Value`. Though, migration to `StateFlow` is tentatively planned for Decompose 4.0.

**Reply by YektaDev** [reply]
Oh, I didn't know that. That's perfect!

**Reply by YektaDev** [reply]
@arkivanov I know this is a bit off-topic, but I was curious about your mention of Decompose 4.0. Is there maybe a public roadmap or a list of potential breaking changes available?

**Reply by arkivanov** [reply]
No roadmap yet, but I'm thinking about creating one. The biggest possible breaking change so far (still unsure about it) is migration from `BackHandlerOwner` to the new [NavigationEventDispatcherOwner](https://developer.android.com/reference/androidx/navigationevent/NavigationEventDispatcherOwner). And there will be a few minor API changes and clean-up.

**Comment by mr-kew** [comment]
Oh I also completely forgot about this question. Well, as I said in the original post, I wanted to do this only because of the ObservableViewModel library. It turned out (perhaps predictably) that it's not a great idea. It basically broke down because the navigation is StateFlow of ViewModels which requires special handling anyway in the ObservableViewModel library. Besides, Decompose pretty much removes the need for the other library and generally makes the whole architecture much cleaner, so I ended up essentially replacing ObservableViewModel with Decompose and never looked back.

After working with Decompose for a bit longer now, I actually pretty like having a Value separate from the StateFlow. I use Value exclusively for the component-view binding and state flow inside the apps logic to essentially cache the flows last state. I think this naming separation works pretty well.

While on the subject though, there are 2 issues with Value that I find a little annoying, but they are not necessarily related to or solvable by using StateFlow.

1. The already mentioned need to always having to implement the View constructor in SwiftUI. It's just tedious and feels unnecessary.
2. Value doesn't do interop (i.e doesn't change `Value<KotlinBoolean>` into `StateValue<Bool>` type) and also it does not work with nullable and closures. That means that I am having to implement a ton of StateValue init overloads just implement the interop and optionals myself and avoid my SwiftUI Views having to deal with a mix of Kotlin & Swift types.

Those are pretty much my only gripes with Value as it is. But the fact that the whole Decompose (not just Value) is so lightweight and easily extensible means that I can just implement the stuff I need myself and it's not an issue. Generally I think the library is pretty much a must on KMP and is very well designed.
